src/main/scala/org/constellation/trust/TrustManager.scala

```diff
@@ -0,0 +1,72 @@
+package org.constellation.trust
+
+import cats.effect.Concurrent
+import cats.implicits._
+import org.constellation.consensus.StoredSnapshot
+import org.constellation.domain.schema.Id
+import org.constellation.domain.trust.TrustData
+import org.constellation.primitives.concurrency.SingleRef
+
+object TrustManager {
+
+  def calculateReputation(snapshots: List[StoredSnapshot]): Map[Id, Double] =
+    snapshots
+      .flatMap(_.checkpointCache.flatMap(_.checkpointBlock.flatMap(_.observations)))
+      .groupBy(o => o.signedObservationData.data.id)
+      .mapValues(_.size.toDouble) // TODO: wkoszycki add conversion List[Observation] -> Score
+
+  def calculateScoringMap(scores: List[TrustData]): Map[Id, Int] =
+    scores.map(_.id).sortBy { _.hex }.zipWithIndex.toMap
+
+  def calculateTrustNodes(scores: List[TrustData], currentNodeId: Id, scoringMap: Map[Id, Int]): List[TrustNode] =
+    scores.map {
+      case TrustData(id, peerScores) =>
+        val selfIdx = scoringMap(id)
+        TrustNode(selfIdx, 0d, 0d, peerScores.map {
+          case (peerId, score) =>
+            TrustEdge(selfIdx, scoringMap(peerId), score, id == currentNodeId)
+        }.toSeq)
+    }
+}
+
+class TrustManager[F[_]: Concurrent](currentNodeId: Id) {
+
+  /**
+    * Reputation based on current node observations
+    */
+  val storedReputation: SingleRef[F, Map[Id, Double]] = SingleRef(Map.empty)
+
+  /**
+    * Reputation based on current node observations with conduction of cluster views
+    */
+  val predictedReputation: SingleRef[F, Map[Id, Double]] = SingleRef(Map.empty)
+
+  import TrustManager._
+
+  def handleTrustScoreUpdate(peerTrustScores: List[TrustData]): F[Unit] =
+    getRecentFinalSnapshots
+      .map(snaps => peerTrustScores :+ TrustData(currentNodeId, TrustManager.calculateReputation(snaps)))
+      .map { scores =>
+        val scoringMap = calculateScoringMap(scores)
+        val idxMap = scoringMap.map(_.swap)
+
+        SelfAvoidingWalk
+          .runWalkFeedbackUpdateSingleNode(
+            scoringMap(currentNodeId),
+            calculateTrustNodes(scores, currentNodeId, scoringMap)
+          )
+          .edges
+          .map { e =>
+            idxMap(e.dst) -> e.trust
+          }
+          .toMap
+      }
+      .flatMap(idMappedScores => predictedReputation.set(idMappedScores))
+
+  def getRecentFinalSnapshots: F[List[StoredSnapshot]] = ??? // TODO: wkoszycki which Snapshots should be loaded ?
```

Comment:
@buckysballs  which snapshots should be considered as `trusted` ?

Reply:
We can get snapshots from the snapshot service. We can still select snapshots with the `HeightIdBasedSnapshotSelector` we just have to change `selectMostRecentCorrectSnapshot` and `selectCorrectRecentSnapshotAtGivenHeight`we just need to change the second maxBy parameter to something like `.maxBy(t => (t._1.height, weightByTrustProposer, t._2.map(_._1._1.hex)))`

where 

`case class RecentSnapshot(hash: String, height: Long, publicReputation: Map[Id, Double] )`

`def weightByTrustProposer(clusterState: Seq[(RecentSnapshot, List[((Id, List[RecentSnapshot]), RecentSnapshot)])]) = {
val proposedTrustViews = clusterState.map(_._1.publicReputation).reduceByKey(_+_)
val clusterStateWeighted = clusterState.map { case (snap, proposals) => 
(snap, proposals.map { k, v => proposedTrustViews(k) )} .sum
}.toMap
clusterStateWeighted
}`

So when a node broadcasts its snapshot, it also sends it's current `publicReputation` along with it, and we weight each snapshot by the sum of all the trust. (sum of occurrences of snapshot hash * sum of trusts proposed)